File: info/1. Алгоритмы/2. Сортировка/19. Быстрая сортировка. Трехчастное разбиение/code sample/Sample1.java
package sample1;

import java.util.Arrays;

public class Sample1 {
    public static void main(String[] args) {

        int[] array = new int[] { 0, 5, -2, 7, 3 };
        quickSort(array);
        System.out.println(Arrays.toString(array));

    }

    public static void quickSort(int[] array) {
        quickSort(array, 0, array.length - 1);
    }

    public static void quickSort(int[] array, int lo, int hi) {
        if (lo >= hi) {
            return;
        }
        int[] part = breakPartition(array, lo, hi);
        quickSort(array, lo, part[0] - 1);
        quickSort(array, part[1] + 1, hi);
    }

    public static int[] breakPartition(int[] array, int lo, int hi) {
        int i = lo + 1;
        int lt = lo;
        int gt = hi;
        int supportElement = array[lo];
        for (; i <= gt;) {
            if (array[i] < supportElement) {
                swap(array, i, lt);
                i += 1;
                lt += 1;
            } else if (array[i] > supportElement) {
                swap(array, i, gt);
                gt -= 1;
            } else {
                i += 1;
            }
        }
        return new int[] { lt, gt };
    }

    public static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

File: info/1. Алгоритмы/2. Сортировка/19. Быстрая сортировка. Трехчастное разбиение/code sample/sample1.py
def quick_sort(sequince, lo_index=None, hi_index=None):
    if lo_index is None:
        lo_index = 0
    if hi_index is None:
        hi_index = len(sequince)-1
    if lo_index >= hi_index:
        return None
    h = partition(sequince, lo_index, hi_index)
    quick_sort(sequince, lo_index, h[0]-1)
    quick_sort(sequince, h[1]+1, hi_index)


def partition(sequince, lo_index, hi_index):
    support_element = sequince[lo_index]
    i = lo_index + 1
    gt = hi_index
    lt = lo_index
    while i <= gt:
        if sequince[i] < support_element:
            sequince[i], sequince[lt] = sequince[lt], sequince[i]
            i += 1
            lt += 1
        elif sequince[i] > support_element:
            sequince[i], sequince[gt] = sequince[gt], sequince[i]
            gt -= 1
        else:
            i += 1
    return [lt, gt]


sequince = [0, 5, -2, 7, 0, 3]

quick_sort(sequince)

print(sequince)
